pow/src/test/java/tech/pegasys/artemis/pow/DepositRequestManagerTest.java

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.pow;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.argThat;
+import static org.mockito.Mockito.atLeast;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static tech.pegasys.artemis.util.config.Constants.ETH1_FOLLOW_DISTANCE;
+
+import com.google.common.primitives.Longs;
+import com.google.common.primitives.UnsignedLong;
+import io.reactivex.Flowable;
+import java.math.BigInteger;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.tuweni.bytes.Bytes;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
+import org.web3j.protocol.core.DefaultBlockParameter;
+import org.web3j.protocol.core.methods.response.EthBlock;
+import org.web3j.protocol.core.methods.response.Log;
+import org.web3j.utils.Numeric;
+import tech.pegasys.artemis.pow.api.DepositEventChannel;
+import tech.pegasys.artemis.pow.contract.DepositContract;
+import tech.pegasys.artemis.pow.event.DepositsFromBlockEvent;
+import tech.pegasys.artemis.util.async.AsyncRunner;
+import tech.pegasys.artemis.util.async.SafeFuture;
+import tech.pegasys.artemis.util.async.StubAsyncRunner;
+
+public class DepositRequestManagerTest {
+
+  private Eth1Provider eth1Provider;
+  private DepositEventChannel depositEventChannel;
+  private DepositContract depositContract;
+  private AsyncRunner asyncRunner;
+
+  private DepositRequestManager depositRequestManager;
+
+  @BeforeEach
+  void setUp() {
+    eth1Provider = mock(Eth1Provider.class);
+    depositEventChannel = mock(DepositEventChannel.class);
+    depositContract = mock(DepositContract.class);
+    asyncRunner = new StubAsyncRunner();
+
+    depositRequestManager =
+        spy(
+            new DepositRequestManager(
+                eth1Provider, asyncRunner, depositEventChannel, depositContract));
+  }
+
+  @Test
+  void restartOnSubscriptionFailure() {
+    mockLatestCanonicalBlockNumber(1);
+    depositRequestManager.start();
+    verify(depositRequestManager, atLeast(2)).start();
+  }
```

Comment:
We want to look for the side effects here, not the call within depositRequestManager.  And it would be better if there was an explicit reason for failure rather than just depending on hitting a random `NullPointerException` later in the chain because a mock isn't fully set up.  So:

```suggestion
@Test
  void restartOnSubscriptionFailure() {
    when(eth1Provider.getLatestBlockFlowable())
        .thenReturn(Flowable.error(new IOException("Node is down")));
    depositRequestManager.start();
    verify(eth1Provider, atLeast(2)).getLatestBlockFlowable();
  }
```
Which exposes an interesting bug because it winds up hitting a `StackOverflowException`.  We probably need a delay before retrying. :)

Reply:
Done.